scarb/src/sources/registry.rs

```diff
@@ -120,28 +128,36 @@ impl<'c> Source for RegistrySource<'c> {
             .iter()
             // NOTE: We filter based on IndexRecords here, to avoid unnecessarily allocating
             //   PackageIds just to abandon them soon after.
+            // NOTE: Technically, RegistryClientCache may already have filtered the records,
+            //   but it is not required to do so, so we do it here again as a safety measure.
             .filter(|record| dependency.version_req.matches(&record.version))
             .map(build_summary_from_index_record)
             .collect())
     }
 
     #[tracing::instrument(level = "trace", skip(self))]
     async fn download(&self, id: PackageId) -> Result<Package> {
-        let is_downloaded = self.client.is_downloaded(id).await;
-
-        ensure!(
-            self.config.network_allowed() || self.client.is_offline() || is_downloaded,
-            "cannot download from `{}` in offline mode",
-            self.source_id
-        );
-
-        if !is_downloaded && !self.client.is_offline() {
-            self.config
-                .ui()
-                .print(Status::new("Downloading", &id.to_string()));
-        }
+        let before_network = {
+            let source_id = self.source_id;
+            let network_allowed = self.config.network_allowed();
+            let ui = self.config.ui();
+            Box::new(move || {
+                ensure!(
```

Comment:
Why isn't this a client's responsibility, to decide if query is possible?

Reply:
Hmm, that's true. My reasoning behind this code was that I wanted to control UI interactions from RegistrySource only rather than smearing it throughout the codebase.

Cargo does smearing and it has a weird `silent` field in clients because it uses clients for waiting for publishing and yet, it doesn't want them to print these “Downloading” statuses along the way.

I'll think about this.

Reply:
Done. I have removed this callback entirely, and replaced it with `Config::online_http` helper and just printing within HTTP's client `download` method. This should be fine even when waiting for downloads, because I realised that logic won't probably need to call download at all, looking at records would be just as fine.